File: hr-simplified/src/main/java/com/example/hrsimplified/service/EmployeeService.java
package com.example.hrsimplified.service;

import com.example.hrsimplified.domain.Employee;
import com.example.hrsimplified.dto.EmployeeDTO;
import com.example.hrsimplified.dto.FilterDTO;
import com.example.hrsimplified.repository.EmployeeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class EmployeeService {

    @Autowired
    private EmployeeRepository employeeRepository;

    public List<Employee> findAll(){
        return employeeRepository.findAll();
    }

    public Employee findById(String id) throws Exception{
        Optional<Employee> checkedEmployee = employeeRepository.findById(Long.valueOf(id));

        if (checkedEmployee.isEmpty()){
            throw new Exception(String.format("Employee with ID %s not found!", id));
        }

        return checkedEmployee.get();
    }

    public Employee findByEmail(String email) throws Exception {
        Optional<Employee> foundEmployee = employeeRepository.findByEmail(email);

        if (foundEmployee.isEmpty()) {
            throw new Exception(String.format("Employee with e-mail %s not found!", email));
        }

        return foundEmployee.get();
    }

    public List<Employee> filter(FilterDTO filter) {
        return employeeRepository.findEmployeeByFilter(
                filter.name(),
                filter.role(),
                filter.department(),
                filter.from(),
                filter.to()
        );
    }

    public Employee save(Employee employee) {
        return employeeRepository.save(employee);
    }

    public Employee create(EmployeeDTO employee) {
        Employee newEmployee = new Employee(employee);

        return this.save(newEmployee);
    }

    public Employee update(EmployeeDTO employee, String id) throws Exception {
        Employee updatedEmployee = new Employee(employee);

        updatedEmployee.setId(Long.valueOf(id));

        return this.save(updatedEmployee);
    }

    public void delete(String id) {
        employeeRepository.deleteById(Long.valueOf(id));
    }
}

File: hr-simplified-frontend/src/app/services/employee.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Employee } from '../interfaces/employee';
import { environment } from '../../environments/environment';
import { EmployeeDTO } from '../interfaces/employee-dto';
import { FilterDTO } from '../interfaces/filter-dto';

@Injectable({
  providedIn: 'root'
})
export class EmployeeService {

  constructor(private http: HttpClient) { }

  findAll(){
    return this.http.get<Employee[]>(`${environment.api}/employee`)
  }

  findById(id: string){
    return this.http.get<Employee>(`${environment.api}/employee/${id}`)
  }

  findByEmail(email: string){
    return this.http.get<Employee>(`${environment.api}/employee/email/${email}`)
  }

  filter(args: FilterDTO) {
    return this.http.post<Employee[]>(`${environment.api}/employee/filter`, args)
  }

  create(args: EmployeeDTO) {
    return this.http.post<Employee>(`${environment.api}/employee`, {
      name: args.name,
      email: args.email,
      role: args.role,
      department: args.department,
      salary: args.salary,
      phone: args.phone,
      hireDate: args.hireDate
    })
  }

  update(args: EmployeeDTO, id: string) {
    return this.http.patch<Employee>(`${environment.api}/employee/${id}`, {
      name: args.name,
      email: args.email,
      role: args.role,
      department: args.department,
      salary: args.salary,
      phone: args.phone,
      hireDate: args.hireDate
    })
  }

  delete(id: string) {
    return this.http.delete<string>(`${environment.api}/employee/${id}`)
  }
}
